### PR TITLE
[BUGFIX] Ensure that all diagnostics for a Checkpoint's children validation definitions appear in error messages

### PR DESCRIPTION
**Pull request by cdkini**
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -109,6 +110,15 @@ class Config:
...
         }
 
+    @override
+    def json(self, **kwargs: Any) -> str:
+        # Necessary override to check that all children validation definitions are added.
+        # JSON encoder will raise an error on the first non-added child.
+        added, errors = self.is_added()
+        if not added:
+            raise CheckpointRelatedResourcesNotAddedError(errors=errors)
+        return super().json(**kwargs)
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: This method is used in the store to serialize - if we don't have this, we check the first validation definition and possibly fail.

Perhaps there is a better way to do this but this allows us to aggregate all errors quite easily.

File: tests/checkpoint/test_checkpoint.py [hunk 2]
```diff
@@ -1002,3 +1002,50 @@ def test_is_added(
...
+    with pytest.raises(CheckpointRelatedResourcesNotAddedError) as e:
+        context.checkpoints.add(
+            gx.Checkpoint(
+                name="my_checkpoint",
+                validation_definitions=[validation_definition_1, validation_definition_2],
+            )
+        )
+
+    errors = e.value.errors
+    assert len(errors) == 4
```
Comment (cdkini) [inline tests/checkpoint/test_checkpoint.py:950]: This would have previously been `len(errors) == 2`

File: tests/checkpoint/test_checkpoint.py [hunk 3]
```diff
@@ -1002,3 +1002,50 @@ def test_is_added(
 
     assert checkpoint_added is is_added
     assert [type(err) for err in errors] == error_list
+
+
+@pytest.mark.unit
```
Comment (cdkini) [inline tests/checkpoint/test_checkpoint.py:916]: Eh is this really a unit test? It's super fast sure but it also didn't feel like e2e

File: tests/checkpoint/test_checkpoint.py [hunk 4]
```diff
@@ -224,111 +225,6 @@ def test_checkpoint_serialization(
         self._assert_valid_uuid(actual["validation_definitions"][0]["id"])
         self._assert_valid_uuid(actual["validation_definitions"][1]["id"])
 
-    @pytest.mark.filesystem
-    def test_checkpoint_filesystem_round_trip_adds_ids(
```
Comment (cdkini) [inline tests/checkpoint/test_checkpoint.py:241]: Round trip should not add ids - we will inform the user of the issue and ask them to persist each item one by one

File: tests/checkpoint/test_checkpoint.py (hunk 3 above)
Comment (tyler-hoffman) [reply tests/checkpoint/test_checkpoint.py:916]: I think it's fine to mark it as a unit test 😄

**Review by tyler-hoffman — APPROVED** [review]

File: great_expectations/checkpoint/checkpoint.py [hunk 5]
```diff
@@ -117,6 +118,19 @@ class Config:
...
+        # Without this, JSON encoder will raise an error on the first non-added child.
+        diagnostics = self.is_added()
+        if not diagnostics.is_added:
+            if not diagnostics.parent_added and diagnostics.children_added:
+                pass  # Checkpoint is not added but all children are - valid serialization state
+            else:
+                diagnostics.raise_for_error()
+
+        return super().json(**kwargs)
+
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: @Kilo59 is there a better way to do this? The JSON encoder will fail on the first error here I want to aggregate all the errors.

File: great_expectations/data_context/store/checkpoint_store.py [hunk 6]
```diff
@@ -63,6 +63,14 @@ def _convert_raw_json_to_object_dict(data: dict) -> dict:
     def serialize(self, value):
         # In order to enable the custom json_encoders in Checkpoint, we need to set `models_as_dict` off  # noqa: E501
         # Ref: https://docs.pydantic.dev/1.10/usage/exporting_models/#serialising-self-reference-or-other-models
+        diagnostics = value.is_added()
+        if self.cloud_mode:
+            # Cloud provides the parent ID
+            diagnostics.raise_for_errors_except_parent()
+        else:
+            # File/ephemeral will add the ID prior to this step
+            diagnostics.raise_for_errors()
```
Comment (cdkini) [inline great_expectations/data_context/store/checkpoint_store.py]: Ugh I don't love this but I rather this than override the public JSON method?

Comment (tyler-hoffman) [reply great_expectations/data_context/store/checkpoint_store.py]: Yeah, I think I like this approach more

**Review by tyler-hoffman — APPROVED** [review]

**Pull request merged**